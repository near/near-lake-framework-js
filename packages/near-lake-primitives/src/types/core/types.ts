export type BlockHeight = number;

export interface StreamerMessage {
    block: BlockView;
    shards: Shard[];
}

export interface BlockView {
    author: string;
    header: BlockHeaderView;
    chunks: ChunkHeader[];
}
export interface BlockHeaderView {
    author: any;
    approvals: (string | null)[];
    blockMerkleRoot: string;
    blockOrdinal: number;
    challengesResult: ChallengeResult[];
    challengesRoot: string;
    chunkHeadersRoot: string;
    chunkMask: boolean[];
    chunkReceiptsRoot: string;
    chunkTxRoot: string;
    chunksIncluded: number;
    epochId: string;
    epochSyncDataHash: string | null;
    gasPrice: string;
    hash: string;
    height: number;
    lastDsFinalBlock: string;
    lastFinalBlock: string;
    latestProtocolVersion: number;
    nextBpHash: string;
    nextEpochId: string;
    outcomeRoot: string;
    prevHash: string;
    prevHeight: number;
    prevStateRoot: string;
    randomValue: string;
    rentPaid: string;
    signature: string;
    timestamp: number;
    timestampNanosec: string;
    totalSupply: string;
    validatorProposals: [];
    validatorReward: string;
}

export interface Shard {
    shardId: number;
    chunk?: ChunkView;
    receiptExecutionOutcomes: ExecutionOutcomeWithReceipt[];
    stateChanges: StateChangeWithCauseView[];
}

export type ValidatorStakeView = {
    accountId: string;
    publicKey: string;
    stake: string;
    validatorStakeStructVersion: string;
};

type ChallengeResult = {
    accountId: string;
    isDoubleSign: boolean;
};


interface ChunkHeader {
    balanceBurnt: number;
    chunkHash: string;
    encodedLength: number;
    encodedMerkleRoot: string;
    gasLimit: number;
    gasUsed: number;
    heightCreated: number;
    heightIncluded: number;
    outcomeRoot: string;
    outgoingReceiptsRoot: string;
    prevBlockHash: string;
    prevStateRoot: string;
    rentPaid: string;
    shardId: number;
    signature: string;
    txRoot: string;
    validatorProposals: ValidatorProposal[];
    validatorReward: string;
}

type ValidatorProposal = {
    accountId: string;
    publicKey: string;
    stake: string;
    validatorStakeStructVersion: string;
};


interface ChunkView {
    author: string;
    header: ChunkHeader;
    receipts: ReceiptView[];
    transactions: IndexerTransactionWithOutcome[];
}

export type ActionReceipt = {
    Action: {
        actions: ActionView[];
        gasPrice: string;
        inputDataIds: string[];
        outputDataReceivers: DataReceiver[];
        signerId: string;
        signerPublicKey: string;
    };
}

export type DataReceipt = {
    Data: {
        data: string;
        dataId: string;
    };
};

type ReceiptEnum =
    | ActionReceipt
    | DataReceipt

type DataReceiver = {
    dataId: string,
    receiverId: string,
};

export type ReceiptView = {
    predecessorId: string;
    receiptId: string;
    receiverId: string;

    receipt: ReceiptEnum;
};

/**
 * `ExecutionStatus` is a simplified representation of the `ExecutionStatusView` from [near-primitives](https://github.com/near/nearcore/tree/master/core/primitives). Represent the execution outcome status for the `Receipt`.
 */
export type ExecutionStatus =
    | { 
    /**
     * Execution succeeded with a value, value is represented by `Uint8Array` and can be anything.
     */
    SuccessValue: Uint8Array }
    | { 
    /**
     * Execution succeeded and a result of the execution is a new `Receipt` with the id.
     */
    SuccessReceiptId: string }
    | { 
    /**
     * Execution failed with an error represented by a `String`.
     */
    Failure: string }
    | 'Postponed'

type ExecutionProof = {
    direction: string;
    hash: string;
};

export type ExecutionOutcomeWithReceipt = {
    executionOutcome: {
        blockHash: string;
        id: string;
        outcome: {
            executorId: string;
            gasBurnt: number;
            logs: string[];
            metadata: {
                gasProfile: string | null;
                version: number;
            };
            receiptIds: string[];
            status: ExecutionStatus;
            tokensBurnt: string;
        };
        proof: ExecutionProof[];
    };
    receipt: ReceiptView;
};

type IndexerTransactionWithOutcome = {
    transaction: Transaction;
    outcome: ExecutionOutcomeWithReceipt;
};


type Transaction = {
    signerId: string;
    publicKey: string;
    nonce: number;
    receiverId: string;
    actions: ActionView[];
    signature: string;
    hash: string;
};

type DeployContractAction = {
    DeployContract: {
        code: string;
    };
};

type FunctionCallAction = {
    FunctionCall: {
        methodName: string;
        args: string;
        gas: number;
        deposit: string;
    };
};

type TransferAction = {
    Transfer: {
        deposit: string;
    };
};

type StakeAction = {
    Stake: {
        stake: number;
        publicKey: string;
    };
};

type AddKeyAction = {
    AddKey: {
        publicKey: string;
        accessKey: AccessKey;
    };
};

interface AccessKey {
    nonce: number;
    permission: string | AccessKeyFunctionCallPermission;
}

interface AccessKeyFunctionCallPermission {
    FunctionCall: {
        allowance: string;
        receiverId: string;
        methodNames: string[];
    }
}

type DeleteKeyAction = {
    DeleteKey: {
        publicKey: string;
    };
};

type DeleteAccountAction = {
    DeleteAccount: {
        beneficiaryId: string;
    };
};

export type DelegateAction = {
  Delegate: {
    delegateAction: {
      senderId: string;
      receiverId: string;
      actions: NonDelegateAction[];
      nonce: number;
      maxBlockHeight: number;
      publicKey: string;
    }
  };
  signature: string;
};

// Delegate actions can not contain other delegate actions
export type NonDelegateAction = 
  | "CreateAccount"
  | DeployContractAction
  | FunctionCallAction
  | TransferAction
  | StakeAction
  | AddKeyAction
  | DeleteKeyAction
  | DeleteAccountAction;

type ActionView =
    | 'CreateAccount'
    | DeployContractAction
    | FunctionCallAction
    | TransferAction
    | StakeAction
    | AddKeyAction
    | DeleteKeyAction
    | DeleteAccountAction
    | DelegateAction;


export type StateChangeWithCauseView = {
    cause: {
        receiptHash: string;
        type: string;
    };
    value: {
        accountId: string;
        keyBase64: string;
        valueBase64: string;
    };
    type: string;
};
