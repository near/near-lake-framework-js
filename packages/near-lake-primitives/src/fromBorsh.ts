import * as borsh from "borsh";
export * from "borsher";

export const fromBorsh = (schema: borsh.Schema, encoded: Uint8Array) =>
  borsh.deserialize(schema, encoded);
