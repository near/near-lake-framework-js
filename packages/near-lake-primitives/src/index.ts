export * from "./types";

export * from "./fromBorsh";
