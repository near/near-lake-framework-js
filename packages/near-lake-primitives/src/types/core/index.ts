export {
    StreamerMessage,
    BlockView,
    BlockHeaderView,
    Shard,
    BlockHeight
} from './types'

