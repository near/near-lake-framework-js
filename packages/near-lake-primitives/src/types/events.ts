
export type Log = {
    log: String;
    relatedReceiptId: String;
}

/**
 * This structure is an ephemeral entity to provide access to the [Events Standard](https://github.com/near/NEPs/blob/master/neps/nep-0297.md) structure and keep data about the related `Receipt` for convenience.
 *
 * #### Interface for Capturing Data About an Event in `handleStreamerMessage()`
 *
 * The interface to capture data about an event has the following arguments:
 *  - `standard`: name of standard, e.g. nep171
 *  - `version`: e.g. 1.0.0
 *  - `event`: type of the event, e.g. `nft_mint`
 *  - `data`: associate event data. Strictly typed for each set {standard, version, event} inside corresponding NEP
 */
export class Event {
    constructor(readonly relatedReceiptId: string, readonly rawEvent: RawEvent) { }

    static fromLog = (log: string): Event => {
        const rawEvent = RawEvent.fromLog(log);
        return new Event('', rawEvent);
    }
}

/**
 * This structure is a copy of the [JSON Events](https://github.com/near/NEPs/blob/master/neps/nep-0297.md) structure representation.
 */
export class RawEvent {
    constructor(readonly event: string, readonly standard: string, readonly version: string, readonly data: JSON | undefined) { }

    static isEvent = (log: string): boolean => {
        return log.startsWith('EVENT_JSON:');
    };

    static fromLog = (log: string): RawEvent => {
        const [event, standard, version, data] = log.split('EVENT_JSON:');

        return new RawEvent(event, standard, version, data ? JSON.parse(data) : undefined);
    };
};

export type Events = {
    events: Event[];
}
