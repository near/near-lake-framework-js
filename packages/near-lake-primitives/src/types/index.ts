export { LakeContext } from './lakeContext';

export { Block } from './block';
export { Event } from './events';
export { Receipt } from './receipts';
export { StateChange } from './stateChanges';
export { Transaction } from './transactions';
export * from './core'


