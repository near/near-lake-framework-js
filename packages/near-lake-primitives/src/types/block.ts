import { Action, Receipt } from './receipts';
import { StreamerMessage, ValidatorStakeView } from './core/types';
import { Transaction } from './transactions';
import { Event, RawEvent, Log } from './events';
import { StateChange } from './stateChanges';

/**
 * The `Block` type is used to represent a block in the NEAR Lake Framework.
 *
 * **Important Notes on `Block`:**
 * - All the entities located on different shards were merged into one single list without differentiation.
 * - `Block` is not the fairest name for this structure either. NEAR Protocol is a sharded blockchain, so its block is actually an ephemeral structure that represents a collection of real blocks called chunks in NEAR Protocol.
 */
export class Block {
    constructor(
        /**
         * Low-level structure for backward compatibility.
         * As implemented in previous versions of [`near-lake-framework`](https://www.npmjs.com/package/near-lake-framework).
         */
        readonly streamerMessage: StreamerMessage,
        private executedReceipts: Receipt[],
        /**
         * Receipts included on the chain but not executed yet marked as “postponed”: they are represented by the same structure `Receipt` (see the corresponding section in this doc for more details).
         */
        readonly postponedReceipts: Receipt[],
        /**
         * List of included `Transactions`, converted into `Receipts`.
         *
         * **_NOTE_:** Heads up! You might want to know about `Transactions` to know where the action chain has begun. Unlike Ethereum, where a Transaction contains everything you may want to know about a particular interaction on  the Ethereum blockchain, Near Protocol because of its asynchronous nature converts a `Transaction` into a `Receipt` before executing it. Thus, On NEAR, `Receipts` are more important for figuring out what happened on-chain as a result of a Transaction signed by a user. Read more about [Transactions on Near](https://nomicon.io/RuntimeSpec/Transactions) here.
         *
         */
        readonly transactions: Transaction[],
        private _actions: Map<string, Action>,
        private _events: Map<string, Event[]>,
        private _stateChanges: StateChange[]) { }

    /**
     * Returns the block hash. A shortcut to get the data from the block header.
     */
    get blockHash(): string {
        return this.header().hash;
    }

    /**
     * Returns the previous block hash. A shortcut to get the data from the block header.
     */
    get prevBlockHash(): string {
        return this.header().prevHash;
    }

    /**
     * Returns the block height. A shortcut to get the data from the block header.
     */
    get blockHeight(): number {
        return this.header().height;
    }

    /**
     * Returns a `BlockHeader` structure of the block
     * See `BlockHeader` structure sections for details.
     */
    header(): BlockHeader {
        return BlockHeader.fromStreamerMessage(this.streamerMessage);
    }

    /**
     * Returns a slice of `Receipts` executed in the block.
     * Basically is a getter for the `executedReceipts` field.
     */
    receipts(): Receipt[] {
        if (this.executedReceipts.length == 0) {
            this.executedReceipts = this.streamerMessage.shards
                .flatMap((shard) => shard.receiptExecutionOutcomes)
                .map((executionReceipt) => Receipt.fromOutcomeWithReceipt(executionReceipt))
        }
        return this.executedReceipts;
    }

    /**
     * Returns an Array of `Actions` executed in the block.
     */
    actions(): Action[] {
        const actions: Action[] = this.streamerMessage.shards
            .flatMap((shard) => shard.receiptExecutionOutcomes)
            .filter((exeuctionOutcomeWithReceipt) => Action.isActionReceipt(exeuctionOutcomeWithReceipt.receipt))
            .map((exeuctionOutcomeWithReceipt) => Action.fromReceiptView(exeuctionOutcomeWithReceipt.receipt))
            .filter((action): action is Action => action !== null)
            .map(action => action)
        return actions
    }

    /**
     * Returns `Events` emitted in the block.
     */
    events(): Event[] {
        const events = this.receipts().flatMap((executedReceipt) => executedReceipt.logs.filter(RawEvent.isEvent).map(RawEvent.fromLog).map((rawEvent) => {
            let event: Event = { relatedReceiptId: executedReceipt.receiptId, rawEvent: rawEvent }
            return event
        }))
        return events
    }

    /**
     * Returns raw logs regardless of the fact that they are standard events or not.
     */
    logs(): Log[] {
        const logs: Log[] = this.receipts().flatMap((executedReceipt) => executedReceipt.logs.map((rawLog) => {
            let log: Log = { relatedReceiptId: executedReceipt.receiptId, log: rawLog }
            return log
        }))
        return logs
    }

    /**
     * Returns an Array of `StateChange` occurred in the block.
     */
    stateChanges(): StateChange[] {
        if (this._stateChanges.length == 0) {
            this._stateChanges = this.streamerMessage.shards
                .flatMap((shard) => shard.stateChanges)
                .map(StateChange.fromStateChangeView)
        }
        return this._stateChanges
    }

    /**
     * Returns `Action` of the provided `receipt_id` from the block if any. Returns `undefined` if there is no corresponding `Action`.
     *
     * This method uses the internal `Block` `action` field which is empty by default and will be filled with the block’s actions on the first call to optimize memory usage.
     *
     * The result is either `Action | undefined` since there might be a request for an `Action` by `receipt_id` from another block, in which case this method will be unable to find the `Action` in the current block. In the other case, the request might be for an `Action` for a `receipt_id` that belongs to a `DataReceipt` where an action does not exist.
     */
    actionByReceiptId(receipt_id: string): Action | undefined {
        if (this._actions.size == 0) {
            this.buildActionsHashmap()
        }
        return this._actions.get(receipt_id);
    }

    /**
     * Returns an Array of Events emitted by `ExecutionOutcome` for the given `receipt_id`. There might be more than one `Event` for the `Receipt` or there might be none of them. In the latter case, this method returns an empty Array.
     */
    eventsByReceiptId(receipt_id: string): Event[] {
        if (this._events.size == 0) {
            this.buildEventsHashmap()
        }
        return this._events.get(receipt_id) || [];
    }

    /**
     * Returns an Array of Events emitted by `ExecutionOutcome` for the given `account_id`. There might be more than one `Event` for the `Receipt` or there might be none of them. In the latter case, this method returns an empty Array.
     */
    eventsByAccountId(account_id: string): Event[] {
        return this.events().filter((event) => {
            const action = this.actionByReceiptId(event.relatedReceiptId)
            return action?.receiverId == account_id || action?.signerId == account_id
        });
    }

    private buildActionsHashmap() {
        const actions = new Map<string, Action>();
        this.actions().forEach(action => {
            actions.set(action.receiptId, action)
        }
        );
        this._actions = actions
    }

    private buildEventsHashmap(): Map<string, Event[]> {
        const events = new Map<string, Event[]>();
        for (const receipt of this.executedReceipts) {
            events.set(receipt.receiptId, receipt.events);
        }
        return events;
    }


    static fromStreamerMessage(streamerMessage: StreamerMessage): Block {
        const block: Block = new Block(streamerMessage, [], [], [], new Map(), new Map(), []);
        return block;
    }

}


/**
 * Replacement for `BlockHeaderView` from [near-primitives](https://github.com/near/nearcore/tree/master/core/primitives). Shrunken and simplified. 
 *
 * **Note:** the original `BlockHeaderView` is still accessible via the `.streamerMessage` attribute.
 */
export class BlockHeader {

    constructor(
        readonly height: number,
        readonly hash: string,
        readonly prevHash: string,
        readonly author: string,
        readonly timestampNanosec: string,
        readonly epochId: string,
        readonly nextEpochId: string,
        readonly gasPrice: string,
        readonly totalSupply: string,
        readonly latestProtocolVersion: number,
        readonly randomValue: string,
        readonly chunksIncluded: number,
        readonly validatorProposals: ValidatorStakeView[]) { }

    static fromStreamerMessage(streamerMessage: StreamerMessage): BlockHeader {
        return new BlockHeader(
            streamerMessage.block.header.height,
            streamerMessage.block.header.hash,
            streamerMessage.block.header.prevHash,
            streamerMessage.block.author,
            streamerMessage.block.header.timestampNanosec,
            streamerMessage.block.header.epochId,
            streamerMessage.block.header.nextEpochId,
            streamerMessage.block.header.gasPrice,
            streamerMessage.block.header.totalSupply,
            streamerMessage.block.header.latestProtocolVersion,
            streamerMessage.block.header.randomValue,
            streamerMessage.block.header.chunksIncluded,
            streamerMessage.block.header.validatorProposals,
        );
    }
}
