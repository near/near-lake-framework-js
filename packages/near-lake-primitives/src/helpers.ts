import { ExecutionStatus, ReceiptStatusFilter } from "./types/core/types";

export function isMatchingReceiverSingle(
  receiverId: string,
  wildcardFilter: string
) {
  if (wildcardFilter === "*") {
    return true;
  }
  const regExp = new RegExp(
    wildcardFilter.replace(/\*/g, "[\\w\\d]+").replace(/\./g, "\\.")
  );
  return regExp.test(receiverId);
}

export function isMatchingReceiver(
  receiverId: string,
  contractFilter: string
): boolean {
  const filters = contractFilter.split(",").map((f) => f.trim());
  return filters.some((f) => isMatchingReceiverSingle(receiverId, f));
}

export function isSuccessfulReceipt(receiptStatus: ExecutionStatus): boolean {
  return (
    receiptStatus.hasOwnProperty("SuccessValue") ||
    receiptStatus.hasOwnProperty("SuccessReceiptId")
  );
}

export function isMatchingReceiptStatus(
  receiptStatus: ExecutionStatus,
  statusFilter: ReceiptStatusFilter
): boolean {
  switch (statusFilter) {
    case "all":
      return true;
    case "onlySuccessful":
      return isSuccessfulReceipt(receiptStatus);
    case "onlyFailed":
      return receiptStatus.hasOwnProperty("Failure");
  }
}
