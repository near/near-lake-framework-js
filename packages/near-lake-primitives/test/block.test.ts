import { readFile } from 'fs/promises';

import { Block } from '../src/types';

describe('Block', () => {
  it('serializes meta transactions', async () => {
    let streamerMessageBuffer = await readFile(`${__dirname}/../../../blocks/105793821.json`);
    let streamerMessage = JSON.parse(streamerMessageBuffer.toString());
    let block = Block.fromStreamerMessage(streamerMessage);

    const actions = block.actionByReceiptId('Dpego7SpsK36PyXjUMrFoSze8ZpNsB9xhb3XJJYtXSix');
    expect(actions?.operations[0]).toMatchSnapshot();
  });
});
