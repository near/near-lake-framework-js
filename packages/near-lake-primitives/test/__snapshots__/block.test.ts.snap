// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block parses event logs 1`] = `
[
  {
    "rawEvent": RawEvent {
      "data": [
        {
          "owner_id": "paras.near",
          "token_ids": [
            "144351:27",
          ],
        },
      ],
      "event": "nft_mint",
      "standard": "nep171",
      "version": "1.0.0",
    },
    "relatedReceiptId": "AVeR4o6MWKYMhjJV8x6ZNk9U9kQYm1qbiN8bBaQQ4rWG",
  },
]
`;

exports[`Block serializes meta transactions 1`] = `
{
  "Delegate": {
    "delegateAction": {
      "actions": [
        {
          "FunctionCall": {
            "args": "eyJhY3Rpb25fbG9ncyI6W3siYWN0aW9uIjoie1wiZnVuY3Rpb25DYWxsXCI6e1wibWV0aG9kTmFtZVwiOlwiZGFpbHlfY2xhaW1cIixcImFyZ3NcIjpcImUzMD1cIixcImdhc1wiOlwiMTAwMDAwMDAwMDAwMDBcIixcImRlcG9zaXRcIjowfSxcImVudW1cIjpcImZ1bmN0aW9uQ2FsbFwifSIsInNpZ25hdHVyZSI6Ijc4Y2Y4ZTAxNTBjZDQ4OWUxOGIzMjU0ZjU4N2U4NGJmOTBmYjA1NWQxYjdmM2UxZTg0YmU5YjdmZmZkMDc0MTYxODVmZmI5ODM0ZDIxNzQyZWYwYmQzZjM1MDI0MjA1NTU3OGYwOWVmYzczOWNiMzZjYzkyOTFjNDFhZWI2NTk4NmZmNDUxOGRhNmM2ODAwNDg3MmFmMWJkZDFmMGVjODZhNWI5ODIzMzAyYmUwMTRkM2ZmNTM0Y2YwNDJkZjAyZjRiNGUzMjNkNWMwNTFmODQ0N2VlZjg4ZTZlMWZiMWI5ZWIyZGI4ODJjMmRjOWU5M2U2MTg2NGFiYjZhNjQwODJ8fDI0MGJkMWQyNTY2MTEwN2E3NTEzOWQ5MDQxNGVjZDc4In1dfQ==",
            "deposit": "0",
            "gas": 10000000000000,
            "methodName": "action_logs",
          },
        },
      ],
      "maxBlockHeight": 105793873,
      "nonce": 105688924000002,
      "publicKey": "ed25519:5T29AvXUFiZusXGfRMYEznCXawhzQKgU5VaWXGgib1Sn",
      "receiverId": "main.arkana.near",
      "senderId": "cuttingnutshell5364174552.u.arkana.near",
    },
    "signature": "ed25519:3dWpDgrNthcsm1dQjBZT2EPg4h3kkrfTM2CvnuMgL8e4aTaUe7mYv1YQ9YCaAL9LKfGTktBWpR3kSxXvCjH2ntyK",
  },
}
`;
